lib/at_cmd/at_cmd.c

```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 			goto next;
 		}
 
-		LOG_DBG("at_cmd_rx %d bytes, %s",
-			bytes_read, log_strdup(item->data));
+		LOG_DBG("at_cmd_rx %d bytes, %s", bytes_read, log_strdup(buf));
 
-		payload_len = get_return_code(item->data, &ret);
-
-		if (ret.state != AT_CMD_NOTIFICATION) {
-			if ((response_buf_len > 0) &&
-			    (response_buf != NULL)) {
-				if (response_buf_len >= payload_len) {
-					memcpy(response_buf, item->data,
-					       payload_len);
-				} else {
-					LOG_ERR("Response buffer not large "
-						"enough");
+		payload_len = get_return_code(buf, &ret);
 
-					ret.code  = -EMSGSIZE;
-				}
-
-				response_buf_len = 0;
-				response_buf     = NULL;
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			ret.callback = notification_handler;
+		}
 
+		/* Allocate a buffer if we need to, verify its size otherwise */
+		if (ret.state == AT_CMD_NOTIFICATION ||
+		    (current_cmd.buf == NULL &&
+		     current_cmd.flags & AT_CMD_BUF_RESP)){
+			ret.buf = k_malloc(payload_len);
```

Comment:
I understand it's needed with current apporach (responses/notifications processed in a system queue), but I just wanted to express my concern, that his may introduce unreliability in notification forwarding in applications that make heavy use of a heap. This might impact modules like Link Controller.

Reply:
Notifications no longer allocate memory